gateway/scaling/function_scaler.go

```diff
@@ -106,3 +139,23 @@ func (f *FunctionScaler) Scale(functionName string) FunctionScaleResult {
 		Duration:  time.Since(start),
 	}
 }
+
+type routine func(attempt int) error
```

Comment:
This is here for logging. I'll have a think whether it could be moved. Logging is not required for every routine a `backoff()` function can run. So this moves the responsibility (as you mentioned) to the routine to decide.